# DINÁMICA MOLECULAR PROTEÍNA-LIGANDO EN GROMACS 
   ## Autor: Emilio Arturo Jesus Paredes Solis
   ## BSc. Ingeniería Electrónica 
   ## Universidad Peruana de Ciencias Aplicadas-UPC

### 1. INSTALACIONES Y DESCARGAS DE PROGRAMAS
### Primero deben descargar la data del PDB Data Bank 
#####  https://www.rcsb.org/    el cristal 6pvz en formato PDB
## Es importante tomar en cuenta:
#### 1.1 Intalar pymol 
#### 1.2 Instalar avogradro ver. 1.2
#### 1.3 Instalar Openbabel(opcional) 
#### 1.4 Instalar pdbfixer 
#### 1.5 Instalar networkx 2.3 // pip install networkx==2.3
#### 1.6 Instalar xmgrace //  sudo apt install grace


#### Finalmente para saber el nombre del ligando abrir el archivo 6pvz.pdb con el editor de texto(gedit) y revisar el nombre de la 4ta columna de HETATM y se podrá observar el nombre. 
#### Otra forma es revisar en el PBD DATA BANK el nombre del ligando o abrir el cristal 6pvz.pdb en pymol y seleccionar el ligando para saber su nombre



## IMPORTANTE: USAR PBBFIXER en este proceso antes de preparar el receptor y el ligando para corregir los átomos perdidos

## Proceso para obtener el receptor con pymol
 
# 2. Se debe preparar ahora con el pymol el receptor eliminando todos los heteroátomos 
### 2.1 Usar comando load 6pvz.pdb  para leer el cristal
### 2.2 Usar comando remove het , para remover heteroàtomos 
### 2.3 Usar comando save receptor.pdb para guardar



## Proceso para obtener el ligando OZM: 
### Usando el comando grep se puede separar el ligando del cristal 
### ejem: grep nombre_del_ligando_MAYUSCULA cristal.pdb > nombre_del ligando_minuscula.pdb 

In [7]:
%%bash
grep OZM 6pvz.pdb > ozm.pdb

## Verificar en pymol el ligando OZM 

In [2]:
%%bash
pymol ozm.pdb

 PyMOL(TM) 2.4.0 - Incentive Product
 Copyright (C) Schrodinger, LLC
 
 This Executable Build integrates and extends Open-Source PyMOL.
 License Expiry date: 01-jan-2021
 Detected OpenGL version 4.6. Shaders available.
 Detected GLSL version 4.60.
 OpenGL graphics engine:
  GL_VENDOR:   NVIDIA Corporation
  GL_RENDERER: GeForce GTX 1050/PCIe/SSE2
  GL_VERSION:  4.6.0 NVIDIA 440.100
 Detected 4 CPU cores.  Enabled multithreaded rendering.
 CmdLoad: "" loaded as "ozm".


## Descargar y descomprimir el campo de fuerza CHARMM36
#### http://mackerell.umaryland.edu/charmm_ff.shtml#gromacs 

In [5]:
%%bash
tar -zxvf charmm36-mar2019.ff.tgz

charmm36-mar2019.ff/
charmm36-mar2019.ff/spc.itp
charmm36-mar2019.ff/ffbonded.itp
charmm36-mar2019.ff/merged.rtp
charmm36-mar2019.ff/nbfix.itp
charmm36-mar2019.ff/gb.itp
charmm36-mar2019.ff/forcefield.doc
charmm36-mar2019.ff/merged.hdb
charmm36-mar2019.ff/ions.itp
charmm36-mar2019.ff/merged.arn
charmm36-mar2019.ff/merged.vsd
charmm36-mar2019.ff/spce.itp
charmm36-mar2019.ff/old_c36_cmap.itp
charmm36-mar2019.ff/ffnonbonded.itp
charmm36-mar2019.ff/atomtypes.atp
charmm36-mar2019.ff/merged.c.tdb
charmm36-mar2019.ff/forcefield.itp
charmm36-mar2019.ff/watermodels.dat
charmm36-mar2019.ff/tip4p.itp
charmm36-mar2019.ff/cmap.itp
charmm36-mar2019.ff/merged.n.tdb
charmm36-mar2019.ff/tip3p.itp
charmm36-mar2019.ff/merged.r2b


## Usando Avogadro versión 1.2 abrir el archivo del ligando ozm.pdb y hacer click en la pestaña Build(Construir), luego seleccionar "añadir hidrógenos" y finalmente guardar el ligando con el formato .mol2 en la carpeta de trabajo. 
### ejem: ozm.mol2


## Usar el comando gmx_mpi pdb2gmx para preparar la topología a partir del receptor preparado, de preferencia ejecutar el siguiente código en un terminal dentro del área de trabajo.  

In [10]:

gmx pdb2gmx -f receptor.pdb -o receptor_processed.gro



SyntaxError: invalid syntax (<ipython-input-10-e684b8835cee>, line 1)

## Ahora ha generado tres archivos nuevos: receptor_processed.gro, topol.top y posre.itp.

<img src="img2.png">

### 1. receptor_processed.gro es un archivo de estructura con formato GROMACS que contiene todos los átomos definidos dentro del campo de fuerza (es decir, se han agregado átomos de H a los aminoácidos en la proteína). 
### 2. El archivo topol.top es la topología del sistema (más sobre esto en un minuto).
### 3. El archivo posre.itp contiene información utilizada para restringir las posiciones de los átomos pesados ​​(revisaremos esto más adelante).

### Observación: En mi caso se obtienen más archivos debido a que estoy trabajando con la cadena A Y B del cristal en el receptor

# TENER MUCHO CUIDADO CON ESTE PASO* 

<img src="img1.png">
Abrir el archivo ozm.mol2 y cambiar al inicio(fila 2) las *** por el nombre del ligando en mayúsculas, en este caso OZM como se observa en la figura


<img src="img3.png">

Aqui presionar CTRL + H para buscar y reemplazar la séptima fila, en este caso OZM201 por OZM en el archivo ozm.mol2  




<img src="img4.png">

Se debe dejar preparado de la siguiente forma finalmente para luego ejecutar un archivo perl que va corregir algunos valores dentro del archivo ozm.mol2

## Descargar script que ordena la data en el siguiente enlace 
### http://www.mdtutorials.com/gmx/complex/Files/sort_mol2_bonds.pl

## Descargar archivo perl y ejecutar script "sort_mol2_bonds.pl"


In [11]:
%%bash
perl sort_mol2_bonds.pl ozm.mol2 ozm_fix.mol2

Found 34 atoms in the molecule, with 36 bonds.


## Crear una cuenta en el servidor CGenFF server 
### https://cgenff.umaryland.edu/ 
### y subir el archivo ozm_fix.mol2 al servidor para generar el archivo .str

<img src="img5.png">
Se puede observar de la siguiente forma la salida




<img src="img6.png">
Se debe copiar todo el código generado y guardar en un editor de texto como el archivo ozm.str


In [16]:
%%bash
python cgenff_charmm2gmx_py3_nx2.py OZM ozm_fix.mol2 ozm.str charmm36-mar2019.ff

NOTE 1: Code tested with Python 3.5.2 and 3.7.3. Your version: 3.7.6 (default, Jan  8 2020, 19:59:22) 
[GCC 7.3.0]

NOTE 2: Code tested with NetworkX 2.3. Your version: 2.3

NOTE 3: Please be sure to use the same version of CGenFF in your simulations that was used during parameter generation:
--Version of CGenFF detected in  ozm.str : 4.0
--Version of CGenFF detected in  charmm36-mar2019.ff/forcefield.doc : 4.1



NOTE 4: To avoid duplicated parameters, do NOT select the 'Include parameters that are already in CGenFF' option when uploading a molecule into CGenFF.
============ DONE ============
Conversion complete.
The molecule topology has been written to ozm.itp
Additional parameters needed by the molecule are written to ozm.prm, which needs to be included in the system .top

PLEASE NOTE: If your topology has lone pairs, you must use GROMACS version 2020 or newer to use 2fd construction
Older GROMACS versions WILL NOT WORK as they do not support 2fd virtual site construction

========

# Construyendo un complejo 
## En el paso anterior se crearon dos archivos: ozm_ini.pdb y ozm.itp(se usaran en breve)
### Desde pdb2gmx, tenemos un archivo llamado "receptor_processed.gro" que contiene la estructura procesada, que cumple con los campos de fuerza de nuestra proteína. También tenemos "ozm_ini.pdb" de cgenff_charmm2gmx.py que tiene todos los átomos H necesarios y coincide con los nombres de los átomos en la topología OZM. Convierta este archivo .pdb a formato .gro con editconf:

In [18]:
%%bash
gmx_mpi editconf -f ozm_ini.pdb -o ozm.gro

Note that major changes are planned in future for editconf, to improve usability and utility.
Read 34 atoms
No velocities found


                     :-) GROMACS - gmx editconf, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
    

## Se abren los archivos receptor_processed.gro, ozm.gro y se crea el archivo complex.gro  

In [19]:
%%bash 
gedit receptor_processed.gro ozm.gro complex.gro 

## Se copia todo el contenido de receptor_processed.gro a complex.gro, y antes de la última fila de éste se copia desde la fila con el primer 1OZM hasta la penúltima fila de ozm.gro a complex.gro 

## 1. En la segunda fila de receptor_processed.gro indica el número de átomos, y este es 5900
## 2. En ozm.gro en la segunda fila indica que tiene 34 átomos 
## 3. En el archivo complex.gro creado se tiene que actualizar la cantidad de átomos, ya que al copiar del archivo receptor_processed.gro tiene 5900, y al agregar los átomos del ligando que son 34 el complejo debería decir que tiene en la segunda fila 5934 átomos.

## IMPORTANTE: Las coordenadas de la última fila de complex.gro es la misma de receptor_processed.gro

# CREANDO LA TOPOLOGÍA

<img src="x2.png">
Ir abajo del archivo topol.top para agregar el la siguiente linea


; Include ligand topology
#include "ozm.itp"


<img src="x1.png">
Ir arriba del archivo topol.top y agregar la siguiente linea como se muestra en la figura


; Include ligand parameters
#include "ozm.prm"


<img src="x3.png">
Finalmente en el archivo topol.top agregar debajo de [ molecules ]  el nombre del ligando, en este caso "OZM" en la última fila como se muestra en la figura

# SOLVATACIÓN
## En este punto, el flujo de trabajo es como cualquier otra simulación de MD. Definiremos la celda unitaria y la llenaremos con agua.

In [21]:
%%bash
gmx_mpi editconf -f complex.gro -o newbox.gro -bt cubic -d 1.0 # ESTE COMANDO CREA LA CAJA DE SOLVATACION, EN ESTE CASO CÚBICA

Note that major changes are planned in future for editconf, to improve usability and utility.
Read 5934 atoms
Volume: 453.431 nm^3, corresponds to roughly 204000 electrons
No velocities found
    system size :  6.492  5.680  5.930 (nm)
    diameter    :  7.268               (nm)
    center      :  1.953 -0.737  1.262 (nm)
    box vectors : 11.759  6.343  6.079 (nm)
    box angles  :  90.00  90.00  90.00 (degrees)
    box volume  : 453.43               (nm^3)
    shift       :  2.681  5.371  3.372 (nm)
new center      :  4.634  4.634  4.634 (nm)
new box vectors :  9.268  9.268  9.268 (nm)
new box angles  :  90.00  90.00  90.00 (degrees)
new box volume  : 796.21               (nm^3)


                     :-) GROMACS - gmx editconf, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
    

In [26]:
%%bash 
vmd newbox.gro   # correrlo de preferencia en linea de comandos para poder ver la caja creada

Info) VMD for LINUXAMD64, version 1.9.4a43 (June 11, 2020)
Info) http://www.ks.uiuc.edu/Research/vmd/                         
Info) Email questions and bug reports to vmd@ks.uiuc.edu           
Info) Please include this reference in published work using VMD:   
Info)    Humphrey, W., Dalke, A. and Schulten, K., `VMD - Visual   
Info)    Molecular Dynamics', J. Molec. Graphics 1996, 14.1, 33-38.
Info) -------------------------------------------------------------
Info) Multithreading available, 4 CPUs detected.
Info)   CPU features: SSE2 AVX AVX2 FMA 
Info) Free system memory: 11GB (69%)
Info) Creating CUDA device pool and initializing hardware...
Info) Detected 1 available CUDA accelerator::
Info) [0] GeForce GTX 1050      5 SM_6.1 1.5 GHz, 3.9GB RAM SP32 KT AE2 ZC
Warning) Detected X11 'Composite' extension: if incorrect display occurs
Warning) try disabling this X server option.  Most OpenGL drivers
Warning) disable stereoscopic display when 'Composite' is enabled.
Info) OpenGL rende

/usr/local/lib/vmd/vmd_LINUXAMD64: /lib/x86_64-linux-gnu/libGL.so.1: no version information available (required by /usr/local/lib/vmd/vmd_LINUXAMD64)


In [27]:
%%bash 
gmx_mpi solvate -cp newbox.gro -cs spc216.gro -p topol.top -o solv.gro # SE REALIZA LA SOLVATACIÓN CON AGUAS Y SE SOBREESCRIBE AUTOMATICAMENTE EN EL ARCHIVO TOPOL.TOP


         based on residue and atom names, since they could not be
         definitively assigned from the information in your input
         files. These guessed numbers might deviate from the mass
         and radius of the atom type. Please check the output
         files if necessary.

NOTE: From version 5.0 gmx solvate uses the Van der Waals radii
from the source below. This means the results may be different
compared to previous GROMACS versions.

++++ PLEASE READ AND CITE THE FOLLOWING REFERENCE ++++
A. Bondi
van der Waals Volumes and Radii
J. Phys. Chem. 68 (1964) pp. 441-451
-------- -------- --- Thank You --- -------- --------

Adding line for 23994 solvent molecules with resname (SOL) to topology file (topol.top)


                     :-) GROMACS - gmx solvate, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
     

# AGREGAR IONES 

## Primero debemos crear con el editor de texto gedit un archivo llamado ions.mdp, copiar el siguiente código y finalmente guardarlo en la carpeta de trabajo (hacer doble click aquí para copiar)

; LINES STARTING WITH ';' ARE COMMENTS
title		    = Minimization	; Title of run

; Parameters describing what to do, when to stop and what to save
integrator	    = steep		; Algorithm (steep = steepest descent minimization)
emtol		    = 1000.0  	; Stop minimization when the maximum force < 10.0 kJ/mol
emstep          = 0.01      ; Energy step size
nsteps		    = 50000	  	; Maximum number of (minimization) steps to perform

; Parameters describing how to find the neighbors of each atom and how to calculate the interactions
nstlist		    = 1		    ; Frequency to update the neighbor list and long range forces
cutoff-scheme   = Verlet
ns_type		    = grid		; Method to determine neighbor list (simple, grid)
rlist		    = 1.0		; Cut-off for making neighbor list (short range forces)
coulombtype	    = cutoff	; Treatment of long range electrostatic interactions
rcoulomb	    = 1.0		; long range electrostatic cut-off
rvdw		    = 1.0		; long range Van der Waals cut-off
pbc             = xyz 		; Periodic Boundary Conditions


## Luego ejecutar el siguiente comando: 

In [29]:
%%bash
gmx_mpi grompp -f ions.mdp -c solv.gro -p topol.top -o ions.tpr

Analysing residue names:
There are:   378    Protein residues
There are:     1      Other residues
There are: 23994      Water residues
Analysing Protein...
Analysing residues not classified as Protein/DNA/RNA/Water and splitting into groups...
This run will generate roughly 6 Mb of data


                      :-) GROMACS - gmx grompp, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
     

In [30]:
%%bash 
gmx_mpi genion -s ions.tpr -o solv_ions.gro -p topol.top -pname NA -nname CL -neutral # correr en linea de comandos


Will try to add 8 NA ions and 0 CL ions.
Select a continuous group of solvent molecules


                      :-) GROMACS - gmx genion, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
     

CalledProcessError: Command 'b'gmx_mpi genion -s ions.tpr -o solv_ions.gro -p topol.top -pname NA -nname CL -neutral\n'' returned non-zero exit status 1.

## Al correr el comando anterior te pregunta que deseas neutralizar, y en nuestro caso sería el sistema,osea el solvente que tendríamos que seleccionar: opción 15

# ENERGÍA DE MINIMIZACIÓN

## Se va crear un nuevo archivo con el editor de texto gedit llamado em.mdp y se va copiar el siguiente script: (hacer doble click aquí para copiar)

; LINES STARTING WITH ';' ARE COMMENTS
title		    = Minimization	; Title of run

; Parameters describing what to do, when to stop and what to save
integrator	    = steep		; Algorithm (steep = steepest descent minimization)
emtol		    = 1000.0  	; Stop minimization when the maximum force < 10.0 kJ/mol
emstep          = 0.01      ; Energy step size
nsteps		    = 50000	  	; Maximum number of (minimization) steps to perform

; Parameters describing how to find the neighbors of each atom and how to calculate the interactions
nstlist		    = 1		        ; Frequency to update the neighbor list and long range forces
cutoff-scheme   = Verlet
ns_type		    = grid		    ; Method to determine neighbor list (simple, grid)
rlist		    = 1.2		    ; Cut-off for making neighbor list (short range forces)
coulombtype	    = PME		    ; Treatment of long range electrostatic interactions
rcoulomb	    = 1.2		    ; long range electrostatic cut-off
vdwtype         = cutoff
vdw-modifier    = force-switch
rvdw-switch     = 1.0
rvdw		    = 1.2		    ; long range Van der Waals cut-off
pbc             = xyz 		    ; Periodic Boundary Conditions
DispCorr        = no



## Luego ejecutar el siguiente comando: 

In [31]:
%%bash
gmx_mpi grompp -f em.mdp -c solv_ions.gro -p topol.top -o em.tpr

Analysing residue names:
There are:   378    Protein residues
There are:     1      Other residues
There are: 23986      Water residues
There are:     8        Ion residues
Analysing Protein...
Analysing residues not classified as Protein/DNA/RNA/Water and splitting into groups...
Analysing residues not classified as Protein/DNA/RNA/Water and splitting into groups...
Calculating fourier grid dimensions for X Y Z
Using a fourier grid of 80x80x80, spacing 0.116 0.116 0.116
This run will generate roughly 6 Mb of data


                      :-) GROMACS - gmx grompp, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
     

In [33]:
%%bash 
gmx_mpi mdrun -v -nb gpu -deffnm em  # version cuda 
# gmx_mpi mdrun -v -deffnm em  # version cpu 


                      :-) GROMACS - gmx mdrun, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
      

# EQUILIBRACIÓN

> gmx_mpi make_ndx -f ozm.gro -o index_ozm.ndx      # (hacer doble click aqui)

 > 0 & ! a H*
 > q
 
 ## Ejecutar linea por en la línea de comandos

<img src="dd.png">

Se va visualizar de esta manera

## Correr el siguiente comando : 
gmx_mpi genrestr -f ozm.gro -n index_ozm.ndx -o posre_ozm.itp -fc 1000 1000 1000


<img src="d1.png">

Se va visualizar de esta manera 
## IMPORTANTE: SELECCIONAR LA OPCIÓN 3

## Luego ir al archivo topol.top y agregar las siguientes lineas: 


; Ligand position restraints
#ifdef POSRES
#include "posre_ozm.itp"
#endif


<img src="d2.png">

Se va visualizar de esta manera y lo guardan

# EQUILIBRACIÓN DE TEMPERATURA
## Antes se va crear un index para separar en dos grupos la proteína y el ligando


## %%bash
## gmx_mpi make_ndx -f em.gro -o index.ndx # correr en linea de comandos

<img src="d3.png">

# > 1 | 13
# > q

## CREAN UN ARCHIVO nvt.mdp con editor de texto y pegan el siguiente script (hacer doble click)

title                   = Protein-ligand complex NVT equilibration 
define                  = -DPOSRES  ; position restrain the protein and ligand
; Run parameters
integrator              = md        ; leap-frog integrator
nsteps                  = 5000     ; 2 * 5000 = 10 ps
dt                      = 0.002     ; 2 fs
; Output control
nstenergy               = 500   ; save energies every 1.0 ps
nstlog                  = 500   ; update log file every 1.0 ps
nstxout-compressed      = 500   ; save coordinates every 1.0 ps
; Bond parameters
continuation            = no        ; first dynamics run
constraint_algorithm    = lincs     ; holonomic constraints 
constraints             = h-bonds   ; bonds to H are constrained 
lincs_iter              = 1         ; accuracy of LINCS
lincs_order             = 4         ; also related to accuracy
; Neighbor searching and vdW
cutoff-scheme           = Verlet
ns_type                 = grid      ; search neighboring grid cells
nstlist                 = 20        ; largely irrelevant with Verlet
rlist                   = 1.2
vdwtype                 = cutoff
vdw-modifier            = force-switch
rvdw-switch             = 1.0
rvdw                    = 1.2       ; short-range van der Waals cutoff (in nm)
; Electrostatics
coulombtype             = PME       ; Particle Mesh Ewald for long-range electrostatics
rcoulomb                = 1.2       ; short-range electrostatic cutoff (in nm)
pme_order               = 4         ; cubic interpolation
fourierspacing          = 0.16      ; grid spacing for FFT
; Temperature coupling
tcoupl                  = V-rescale                     ; modified Berendsen thermostat
tc-grps                 = Protein_OZM Water_and_ions    ; two coupling groups - more accurate
tau_t                   = 0.1   0.1                     ; time constant, in ps
ref_t                   = 300   300                     ; reference temperature, one for each group, in K
; Pressure coupling
pcoupl                  = no        ; no pressure coupling in NVT
; Periodic boundary conditions
pbc                     = xyz       ; 3-D PBC
; Dispersion correction is not used for proteins with the C36 additive FF
DispCorr                = no 
; Velocity generation
gen_vel                 = yes       ; assign velocities from Maxwell distribution
gen_temp                = 300       ; temperature for Maxwell distribution
gen_seed                = -1        ; generate a random seed



## El script trabaja en 10 ps por motivo de tiempo

In [34]:
%%bash 
gmx_mpi grompp -f nvt.mdp -c em.gro -r em.gro -p topol.top -n index.ndx -o nvt.tpr

turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
Determining Verlet buffer for a tolerance of 0.005 kJ/mol/ps at 300 K
Calculated rlist for 1x1 atom pair-list as 1.293 nm, buffer size 0.093 nm
Set rlist, assuming 4x4 atom pair-list, to 1.221 nm, buffer size 0.021 nm
Note that mdrun will redetermine rlist based on the actual pair-list setup
Calculating fourier grid dimensions for X Y Z
Using a fourier grid of 60x60x60, spacing 0.154 0.154 0.154
This run will generate roughly 10 Mb of data


                      :-) GROMACS - gmx grompp, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
     

In [36]:
%%bash 
gmx_mpi mdrun -v -nb gpu -deffnm nvt  # ejecutar en linea de comandos version CUDA
# gmx mdrun -deffnm nvt  # ejecutar en linea de comandos version CPU


                      :-) GROMACS - gmx mdrun, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
      

CalledProcessError: Command 'b'gmx_mpi mdrun -v -nb gpu -deffnm nvt\n'' returned non-zero exit status 139.

<img src="d5.png">


# EQUILIBRACIÓN NPT 

## Crear un archivo npt.mdp con el editor de texto y copiar el script(hacer click aqui dos veces y pegar)

title                   = Protein-ligand complex NPT equilibration 
define                  = -DPOSRES  ; position restrain the protein and ligand
; Run parameters
integrator              = md        ; leap-frog integrator
nsteps                  = 5000     ; 2 * 5000 = 10 ps
dt                      = 0.002     ; 2 fs
; Output control
nstenergy               = 500       ; save energies every 1.0 ps
nstlog                  = 500       ; update log file every 1.0 ps
nstxout-compressed      = 500       ; save coordinates every 1.0 ps
; Bond parameters
continuation            = yes       ; continuing from NVT 
constraint_algorithm    = lincs     ; holonomic constraints 
constraints             = h-bonds   ; bonds to H are constrained 
lincs_iter              = 1         ; accuracy of LINCS
lincs_order             = 4         ; also related to accuracy
; Neighbor searching and vdW
cutoff-scheme           = Verlet
ns_type                 = grid      ; search neighboring grid cells
nstlist                 = 20        ; largely irrelevant with Verlet
rlist                   = 1.2
vdwtype                 = cutoff
vdw-modifier            = force-switch
rvdw-switch             = 1.0
rvdw                    = 1.2       ; short-range van der Waals cutoff (in nm)
; Electrostatics
coulombtype             = PME       ; Particle Mesh Ewald for long-range electrostatics
rcoulomb                = 1.2
pme_order               = 4         ; cubic interpolation
fourierspacing          = 0.16      ; grid spacing for FFT
; Temperature coupling
tcoupl                  = V-rescale                     ; modified Berendsen thermostat
tc-grps                 = Protein_OZM Water_and_ions    ; two coupling groups - more accurate
tau_t                   = 0.1   0.1                     ; time constant, in ps
ref_t                   = 300   300                     ; reference temperature, one for each group, in K
; Pressure coupling
pcoupl                  = Berendsen                     ; pressure coupling is on for NPT
pcoupltype              = isotropic                     ; uniform scaling of box vectors
tau_p                   = 2.0                           ; time constant, in ps
ref_p                   = 1.0                           ; reference pressure, in bar
compressibility         = 4.5e-5                        ; isothermal compressibility of water, bar^-1
refcoord_scaling        = com
; Periodic boundary conditions
pbc                     = xyz       ; 3-D PBC
; Dispersion correction is not used for proteins with the C36 additive FF
DispCorr                = no 
; Velocity generation
gen_vel                 = no        ; velocity generation off after NVT 

In [37]:
%%bash 
gmx_mpi grompp -f npt.mdp -c nvt.gro -t nvt.cpt -r nvt.gro -p topol.top -n index.ndx -o npt.tpr

turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
Determining Verlet buffer for a tolerance of 0.005 kJ/mol/ps at 300 K
Calculated rlist for 1x1 atom pair-list as 1.293 nm, buffer size 0.093 nm
Set rlist, assuming 4x4 atom pair-list, to 1.221 nm, buffer size 0.021 nm
Note that mdrun will redetermine rlist based on the actual pair-list setup
Calculating fourier grid dimensions for X Y Z
Using a fourier grid of 60x60x60, spacing 0.154 0.154 0.154
This run will generate roughly 10 Mb of data


                      :-) GROMACS - gmx grompp, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
     


# gmx_mpi mdrun -v -nb gpu -deffnm npt  (correr en linea de comandos)  ó
# gmx  mdrun  -deffnm npt  (correr en linea de comandos versión cpu)

<img src="m1.png">

# SIMULACIÓN

## Una vez completadas las dos fases de equilibrio, el sistema ahora está bien equilibrado a la temperatura y presión deseadas. Ahora estamos listos para liberar las restricciones de posición y ejecutar MD de producción para la recopilación de datos. El proceso es como lo hemos visto antes, ya que haremos uso del archivo de punto de control (que en este caso ahora contiene información de acoplamiento de presión de preservación) para grompp. Ejecutaremos una simulación MD de 100 ps, cuyo script se puede encontrar aquí.



title                   = Protein-ligand complex MD simulation 
; Run parameters
integrator              = md        ; leap-frog integrator
nsteps                  = 50000   ; 2 * 50000 = 100 ps 
dt                      = 0.002     ; 2 fs
; Output control
nstenergy               = 5000      ; save energies every 10.0 ps
nstlog                  = 5000      ; update log file every 10.0 ps
nstxout-compressed      = 5000      ; save coordinates every 10.0 ps
; Bond parameters
continuation            = yes       ; continuing from NPT 
constraint_algorithm    = lincs     ; holonomic constraints 
constraints             = h-bonds   ; bonds to H are constrained
lincs_iter              = 1         ; accuracy of LINCS
lincs_order             = 4         ; also related to accuracy
; Neighbor searching and vdW
cutoff-scheme           = Verlet
ns_type                 = grid      ; search neighboring grid cells
nstlist                 = 20        ; largely irrelevant with Verlet
rlist                   = 1.2
vdwtype                 = cutoff
vdw-modifier            = force-switch
rvdw-switch             = 1.0
rvdw                    = 1.2       ; short-range van der Waals cutoff (in nm)
; Electrostatics
coulombtype             = PME       ; Particle Mesh Ewald for long-range electrostatics
rcoulomb                = 1.2
pme_order               = 4         ; cubic interpolation
fourierspacing          = 0.16      ; grid spacing for FFT
; Temperature coupling
tcoupl                  = V-rescale                     ; modified Berendsen thermostat
tc-grps                 = Protein_OZM Water_and_ions    ; two coupling groups - more accurate
tau_t                   = 0.1   0.1                     ; time constant, in ps
ref_t                   = 300   300                     ; reference temperature, one for each group, in K
; Pressure coupling 
pcoupl                  = Parrinello-Rahman             ; pressure coupling is on for NPT
pcoupltype              = isotropic                     ; uniform scaling of box vectors
tau_p                   = 2.0                           ; time constant, in ps
ref_p                   = 1.0                           ; reference pressure, in bar
compressibility         = 4.5e-5                        ; isothermal compressibility of water, bar^-1
; Periodic boundary conditions
pbc                     = xyz       ; 3-D PBC
; Dispersion correction is not used for proteins with the C36 additive FF
DispCorr                = no 
; Velocity generation
gen_vel                 = no        ; continuing from NPT equilibration 


In [38]:
%%bash 
gmx_mpi grompp -f md.mdp -c npt.gro -t npt.cpt -p topol.top -n index.ndx -o md_0_10.tpr

turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
turning H bonds into constraints...
Determining Verlet buffer for a tolerance of 0.005 kJ/mol/ps at 300 K
Calculated rlist for 1x1 atom pair-list as 1.295 nm, buffer size 0.095 nm
Set rlist, assuming 4x4 atom pair-list, to 1.222 nm, buffer size 0.022 nm
Note that mdrun will redetermine rlist based on the actual pair-list setup
Calculating fourier grid dimensions for X Y Z
Using a fourier grid of 60x60x60, spacing 0.152 0.152 0.152
This run will generate roughly 10 Mb of data


                      :-) GROMACS - gmx grompp, 2020.2 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar      Christian Blau   Viacheslav Bolnykh     Kevin Boyd    
 Aldert van Buuren   Rudi van Drunen     Anton Feenstra       Alan Gray     
  Gerrit Groenhof     Anca Hamuraru    Vincent Hindriksen  M. Eric Irrgang  
  Aleksei Iupinov   Christoph Junghans     Joe Jordan     Dimitrios Karkoulis
    Peter Kasson        Jiri Kraus      Carsten Kutzner      Per Larsson    
  Justin A. Lemkul    Viveca Lindahl    Magnus Lundborg     Erik Marklund   
    Pascal Merz     Pieter Meulenhoff    Teemu Murtola       Szilard Pall   
    Sander Pronk      Roland Schulz      Michael Shirts    Alexey Shvetsov  
   Alfons Sijbers     Peter Tieleman      Jon Vincent      Teemu Virolainen 
 Christian Wennberg    Maarten Wolf      Artem Zhmurov   
                           and the project leaders:
     

## ejecutar el siguiente comando : gmx_mpi mdrun -v -nb gpu -deffnm md_0_10   

## ó ejecutar gmx mdrun -deffnm md_0_10(versión cpu)

<img src="m2.png">

# FINALMENTE PARA VISUALIZAR NUESTRA SIMULACIÒN EJECUTAR EL SIGUIENTE COMANDO: 

gmx_mpi trjconv -s md_0_10.tpr -f md_0_10.xtc -o md_0_10_center.xtc -center -pbc mol -ur compact 

## Es para centrar, se debe seleccionar la opción 1 (protein) y la opción 13(ligand) 


# FINALMENTE LUEGO DE CENTRAR, SE PROCEDE A VISUALIZAR CON VMD LOS RESULTADOS DE LA DINÁMICA MOLECULAR PROTEÌNA-LIGANDO

In [40]:
%%bash 
#vmd md_0_10.gro md_0_10.xtc

<img src="f.png">

# OBSERVACIONES: 

## 1. Este sistema a sido preparado con Gromacs 2020.2 versión gpu, si usted no posee gpu reemplace en los comando gmx_mpi solo por gmx 
## 2. Tener cuidado al ejecutar los comandos en NVT, NPT Y MD ya que la versión cpu tiene comando un poco diferentes y se han adjuntado según usted vea conveniente. 

# ANÁLISIS DE RESULTADOS

In [2]:
#gmx_mpi trjconv -s md_0_20.tpr -f md_0_20.xtc -o md_0_20_center.xtc -center -pbc mol -ur compact
# 1 
# 0
# Centrando las trayectorias


In [3]:
#gmx_mpi rms -s md_0_20.tpr -f md_0_20_center.xtc -o rmsd_complex.xvg -tu ns
#1
#13
#EOF
# RMSD plot complex

In [ ]:
#gmx_mpi rms -s md_0_20.tpr -f md_0_20_center.xtc -o rmsd_protein.xvg -tu ns
#1
#1
# RMSD plot protein

In [4]:
#gmx_mpi rms -s md_0_20.tpr -f md_0_20_center.xtc -o rmsd_ligand.xvg -tu ns
#13
#13
# RMSD plot ligand

In [5]:
#xmgrace *.xvg
# RMSD plot complex, protein and ligand

In [6]:
#gmx_mpi rmsf -s md_0_20.tpr -f md_0_20_center.xtc -res -o rmsf.xvg
#1
# RMSF plot

In [7]:
#xmgrace rmsf.xvg

# EXTENDIENDO LA DINÁMICA

In [8]:
#gmx_mpi convert-tpr -s md_0_10.tpr -extend 1000 -o md_0_20.tpr

In [9]:
#gmx mdrun -v -nb gpu -deffnm md_0_20

In [10]:
#gmx_mpi trjcat -f md_0_1.xtc md_0_30.xtc -o md_0_final.xtc -cat

# ¡¡MUCHAS GRACIAS!!

# TERMINAMOS NUESTRA DINÁMICA MOLECULAR EN GROMACS